### ISSUE TITLE
Terraform language server is slow to update, high CPU utilization

### ISSUE THREAD

**Issue by cryptk** [hashicorp/terraform-ls#520]
### Server Version

```
0.16.3
```

### Terraform Version

```
Terraform v0.14.9
```

### Client Version

```
VSCode with hashicorp.terraform extension v2.11.0
```

### Terraform Configuration Files

This is an internal repository

### Expected Behavior
Terraform-ls should run quickly and performantly

### Actual Behavior
Terraform runs slowly and uses a large amount of CPU.  I stepped through the commits to find the one that started the performance regression and it appears that https://github.com/hashicorp/terraform-ls/commit/1de4bb5178320518e311a523324acaedb3519ede is the commit where I began to observe the performance regression.  Here is a top25 from a CPU profile of terraform-ls while it was experiencing the performance regression:

```
File: terraform-ls
Type: cpu
Time: May 19, 2021 at 11:49am (CDT)
Duration: 22.37s, Total samples = 13.64s (60.98%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top25
Showing nodes accounting for 10.29s, 75.44% of 13.64s total
Dropped 301 nodes (cum <= 0.07s)
Showing top 25 nodes out of 120
      flat  flat%   sum%        cum   cum%
     1.45s 10.63% 10.63%      1.98s 14.52%  runtime.mapiternext
     1.14s  8.36% 18.99%     12.42s 91.06%  github.com/mitchellh/reflectwalk.walk
     0.94s  6.89% 25.88%      3.42s 25.07%  github.com/mitchellh/copystructure.(*walker).Exit
     0.77s  5.65% 31.52%     12.42s 91.06%  github.com/mitchellh/reflectwalk.walkStruct
     0.77s  5.65% 37.17%      1.89s 13.86%  runtime.mallocgc
     0.61s  4.47% 41.64%      0.73s  5.35%  runtime.(*itabTableType).find
     0.49s  3.59% 45.23%      1.22s  8.94%  runtime.getitab
     0.39s  2.86% 48.09%      1.61s 11.80%  runtime.assertE2I2
     0.39s  2.86% 50.95%      0.39s  2.86%  runtime.nextFreeFast (inline)
     0.33s  2.42% 53.37%      2.15s 15.76%  runtime.newobject
     0.30s  2.20% 55.57%      1.33s  9.75%  runtime.mapiterinit
     0.29s  2.13% 57.70%     12.19s 89.37%  github.com/mitchellh/reflectwalk.walkSlice
     0.27s  1.98% 59.68%      0.27s  1.98%  runtime.isEmpty (inline)
     0.26s  1.91% 61.58%      0.26s  1.91%  runtime.duffcopy
     0.25s  1.83% 63.42%      0.25s  1.83%  runtime.add (inline)
     0.23s  1.69% 65.10%      0.25s  1.83%  runtime.newdefer
     0.20s  1.47% 66.57%      0.21s  1.54%  runtime.freedefer
     0.20s  1.47% 68.04%      0.20s  1.47%  runtime.memmove
     0.18s  1.32% 69.35%      0.53s  3.89%  runtime.deferproc
     0.18s  1.32% 70.67%      0.26s  1.91%  runtime.scanobject
     0.15s  1.10% 71.77%      0.27s  1.98%  runtime.heapBitsSetType
     0.13s  0.95% 72.73%      0.13s  0.95%  github.com/mitchellh/copystructure.(*walker).Enter
     0.13s  0.95% 73.68%      0.13s  0.95%  runtime.futex
     0.12s  0.88% 74.56%      0.12s  0.88%  github.com/apparentlymart/go-textseg/v13/textseg.ScanGraphemeClusters
     0.12s  0.88% 75.44%      0.95s  6.96%  reflect.(*structType).Field
```

If a terraform-ls developer would like a copy of the prof file, let me know.

### Steps to Reproduce
Edit TF files and note the slow performance

**Comment by radeksimko** [comment]
Hi @cryptk 
There is a known issue causing high CPU utilization in 0.16.0+ versions described in https://github.com/hashicorp/terraform-ls/issues/509

Do you mind [collecting CPU profile](https://github.com/hashicorp/terraform-ls/blob/main/docs/TROUBLESHOOTING.md#cpu-profiling) and providing it back (e.g. by uploading the file to your next comment) just so we can confirm the root cause is that same and that it will also be resolved via https://github.com/hashicorp/terraform-ls/pull/513?

Thank you!

**Comment by cryptk** [comment]
@radeksimko sure thing! Here is the CPU profile.  I had to rename it to a .gz to have it be a github supported format.

[terraform-ls-cpu.prof.gz](https://github.com/hashicorp/terraform-ls/files/6510561/terraform-ls-cpu.prof.gz)

**Comment by radeksimko** [comment]
The CPU profile looks very similar to the other ones I have collected, so I believe that this should be addressed in 0.17.1 which brings the CPU usage down via https://github.com/hashicorp/terraform-ls/pull/513

@cryptk Can you install that new version and confirm, please?

**Comment by radeksimko** [comment]
I'm going to assume this was resolved as part of #513 but feel free to reach out by opening a new issue and attaching a fresh CPU profile if you keep running into problems.

Thanks.
